### PR TITLE
Do not serve certificate content for Non-SSL routes

### PR DESCRIPTION
**Pull request by pecameron**
By default, when a host does not resolve to a route in a HTTPS or tls
sni request, the default cert is returned to the caller as part of the
503 response. This exposes the default cert and may pose security
concerns. Haproxy strict-sni option to bind suppresses use of the
default cert.

This adds a new environment variable to the router deployment controller,
ROUTER_STRICT_SNI, to control bind processing. When set to "true" or
"TRUE", "strict-sni" is added to the bind. Default is "false".

oc adm router --strict-sni sets ROUTER_STRICT_SNI="true"

bug 1369865
https://bugzilla.redhat.com/show_bug.cgi?id=1369865

### CODE REVIEW

**Comment by pecameron** [comment]
@rajatchopra @knobunc First part, haproxy-template.conf change. PTAL

**Comment by pecameron** [comment]
@knobunc @rajatchopra @JacobTanenbaum Added command option PTAL

File: images/router/haproxy/conf/haproxy-config.template [hunk 1]
```diff
@@ -204,7 +204,7 @@ backend be_sni
 
 frontend fe_sni
   # terminate ssl on edge
-  bind 127.0.0.1:{{env "ROUTER_SERVICE_SNI_PORT" "10444"}} ssl no-sslv3 {{ if gt (len .DefaultCertificate) 0 }}crt {{.DefaultCertificate}}{{ else }}crt /var/lib/haproxy/conf/default_pub_keys.pem{{ end }} crt-list /var/lib/haproxy/conf/cert_config.map accept-proxy
+  bind 127.0.0.1:{{env "ROUTER_SERVICE_SNI_PORT" "10444"}} ssl no-sslv3 {{ if matchPattern "true|TRUE" (env "ROUTER_STRICT_SNI" "") -}} strict-sni {{ end }}{{ if gt (len .DefaultCertificate) 0 }}crt {{.DefaultCertificate}}{{ else }}crt /var/lib/haproxy/conf/default_pub_keys.pem{{ end }} crt-list /var/lib/haproxy/conf/cert_config.map accept-proxy
```
Comment (knobunc) [inline images/router/haproxy/conf/haproxy-config.template]: It looks right to me, but I'd like to see some whitespace so I can follow what we are doing a little more clearly:

bind 127.0.0.1:{{env "ROUTER_SERVICE_SNI_PORT" "10444"}} ssl no-sslv3 
  {{- if matchPattern "true|TRUE" (env "ROUTER_STRICT_SNI" "") }} strict-sni {{ end }}
  {{- if gt (len .DefaultCertificate) 0 }} crt {{.DefaultCertificate}} {{ else }} crt /var/lib/haproxy/conf/default_pub_keys.pem {{ end }}
  {{- ""}} crt-list /var/lib/haproxy/conf/cert_config.map accept-proxy

**Comment by knobunc** [comment]
[test]

**Comment by knobunc** [comment]
@openshift/networking PTAL

**Comment by pecameron** [comment]
@knobunc PTAL

File: test/cmd/router.sh [hunk 2]
```diff
@@ -44,6 +44,8 @@ os::cmd::expect_failure_and_text 'oadm router --dry-run --host-network=false --h
 os::cmd::expect_success_and_text 'oadm router --dry-run --host-network=false --host-ports=false --max-connections=14583 -o yaml' '14583'
 # ciphers 
 os::cmd::expect_success_and_text 'oadm router --dry-run --host-network=false --host-ports=false --ciphers=modern -o yaml' 'modern'
+# strict-sni
+os::cmd::expect_success_and_text 'oadm router --dry-run --host-network=false --host-ports=false --strict-sni -o yaml' 'ROUTER_STRICT_SNI'
```
Comment (knobunc) [inline test/cmd/router.sh]: Is this check sufficient?  You always set the variable, but to true or false as appropriate.

File: images/router/haproxy/conf/haproxy-config.template [hunk 3]
```diff
@@ -204,7 +204,10 @@ backend be_sni
 
 frontend fe_sni
   # terminate ssl on edge
-  bind 127.0.0.1:{{env "ROUTER_SERVICE_SNI_PORT" "10444"}} ssl no-sslv3 {{ if gt (len .DefaultCertificate) 0 }}crt {{.DefaultCertificate}}{{ else }}crt /var/lib/haproxy/conf/default_pub_keys.pem{{ end }} crt-list /var/lib/haproxy/conf/cert_config.map accept-proxy
+  bind 127.0.0.1:{{env "ROUTER_SERVICE_SNI_PORT" "10444"}} ssl no-sslv3
+    {{- if matchPattern "true|TRUE" (env "ROUTER_STRICT_SNI" "") }} strict-sni {{ end }}
```
Comment (pravisankar) [inline images/router/haproxy/conf/haproxy-config.template]: Will this end up putting 'strict-sni' on the same line as 'bind 127.0.0.1'? (Assuming {{-' only  removes leading spaces)

Comment (rajatchopra) [inline images/router/haproxy/conf/haproxy-config.template]: Will this collapse on the previous line? not sure how the white-space collapse works.
All else looks good.

**Comment by knobunc** [comment]
[test]

**Comment by pravisankar** [comment]
@pecameron Can you split this change into 2 commits. One commit for the actual change and another commit for the auto generated stuff that way it will be cleaner.
I added one other comment, otherwise change looks good.

**Comment by knobunc** [comment]
@pravisankar is that safe?  Won't one commit fail tests because the autogen stuff isn't current?  Thus it will break bisect?  Or are you asking that we do it for the PR only and then squash before it is merged?

File: test/cmd/router.sh (hunk 2 above)
Comment (JacobTanenbaum) [inline test/cmd/router.sh]: I think you should check the value

**Comment by pravisankar** [comment]
On Wed, Jun 14, 2017 at 12:08 PM, Ben Bennett <notifications@github.com>
wrote:

> @pravisankar <https://github.com/pravisankar> is that safe? Won't one
> commit fail tests because the autogen stuff isn't current? Thus it will
> break bisect? Or are you asking that we do it for the PR only and then
> squash before it is merged?
>
No, it won't break. Test/Merge is performed on the entire changes rebased
on master. This was suggested on my PR long time back and I have seen
others following this convention (just grep for 'generated' in 'git log'
output).

**Comment by knobunc** [comment]
[merge]

**Comment by pecameron** [comment]
Yes it does put strict-sni on the bind line (I verified this)

**Pull request merged**